src/txn/txn_ckpt.c

```diff
@@ -239,22 +239,67 @@ int
 __wt_checkpoint_get_handles(WT_SESSION_IMPL *session, const char *cfg[])
 {
 	WT_BTREE *btree;
+	WT_CURSOR *meta_cursor;
 	WT_DECL_RET;
 	const char *name;
+	bool metadata_race;
+
+	btree = S2BT(session);
 
 	/* Should not be called with anything other than a file object. */
 	WT_ASSERT(session, session->dhandle->checkpoint == NULL);
 	WT_ASSERT(session, WT_PREFIX_MATCH(session->dhandle->name, "file:"));
 
 	/* Skip files that are never involved in a checkpoint. */
-	if (F_ISSET(S2BT(session), WT_BTREE_NO_CHECKPOINT))
+	if (F_ISSET(btree, WT_BTREE_NO_CHECKPOINT))
 		return (0);
 
-	/* Make sure there is space for the next entry. */
-	WT_RET(__wt_realloc_def(session, &session->ckpt_handle_allocated,
-	    session->ckpt_handle_next + 1, &session->ckpt_handle));
+	/*
+	 * We may have raced between starting the checkpoint transaction and
+	 * some operation completing on the handle that updated the metadata
+	 * (e.g., closing a bulk load cursor).  All such operations either have
+	 * exclusive access to the handle or hold the schema lock.  We are now
+	 * holding the schema lock and have an open btree handle, so if we
+	 * can't update the metadata, then there has been some state change
+	 * invisible to the checkpoint transaction.  Skip checkpointing such
+	 * files: they must have a recent durable point.
+	 */
+	if (!WT_IS_METADATA(session->dhandle)) {
+		WT_ASSERT(session, !F_ISSET(&session->txn, WT_TXN_ERROR));
+		WT_RET(__wt_metadata_cursor(session, &meta_cursor));
+		meta_cursor->set_key(meta_cursor, session->dhandle->name);
+		ret = __wt_curfile_update_check(meta_cursor);
+		if (ret == WT_ROLLBACK) {
+			metadata_race = true;
+			ret = 0;
+		} else
+			metadata_race = false;
+		WT_TRET(__wt_metadata_cursor_release(session, &meta_cursor));
+		WT_RET(ret);
+		if (metadata_race) {
+			__wt_verbose(session, WT_VERB_CHECKPOINT,
+			    "skipped checkpoint of %s with metadata conflict",
+			    session->dhandle->name);
+			F_CLR(&session->txn, WT_TXN_ERROR);
+			F_SET(btree, WT_BTREE_SKIP_CKPT);
+			return (0);
+		}
+	}
```

Comment:
I'm not clear on what "a recent durable point" means, exactly; since the durable point must precede the transaction's durable point, I don't see any problems. with one question: presumably the durable point was a default checkpoint, what if this checkpoint is named?